### PR TITLE
Add tests of LMDBDataset

### PR DESCRIPTION
**Pull request by frreiss**
This pull requests adds regression tests for the `LMDBDatasetOp` class. I used the most recent version of the `DatasetOpsTestBaseV2` test harness class as of this morning, including the new macros like `ITERATOR_GET_NEXT_TEST_P`.

I also refactored the contents of `lmdb_dataset_op.cc` into `.cc` and `.h` files so that the test cases could create instances of `LMDBDatasetOp`.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Can one of the admins verify this patch?

File: tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+  std::vector<Tensor> ret;
+  for (string str : strs) {
+    ret.push_back(CreateTensor<string>(TensorShape({}), {str}));
+  }
+  return ret;
+}
+
+// The tensors we expect to see each time we read through the input
+// data file.
+const std::vector<Tensor> kFileOutput = StringsToTensors({
```
Comment (feihugis) [inline tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc]: The util function [CreateTensors()](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/dataset_test_base.h#L95) can be used here, for example
```C++
const std::vector<Tensor> kFileOutput = CreateTensors<string>(TensorShape({}), {{"0"}, {"a"}, ..., {"j"}};
```

Comment (frreiss) [reply tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc]: Thanks for pointing that out! I have switched to using `CreateTensors` in commit 03140b3.

**Review by jsimsa — DISMISSED** [review]

**Comment by gbaned** [comment]
@frreiss Could you please check reviewer comments and keep us posted. Thanks!

**Comment by frreiss** [comment]
Sorry for the delay in getting back to this. I've been in back to back meetings for the last three days. Looking into the build file issue now.

**Review by jsimsa — DISMISSED** [review]

**Comment by jsimsa** [comment]
Internal tests fails with:

```
third_party/tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc:15:10: error: module //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/io/path.h'
see http://go/cpp-features#layering_check; to fix run:
build_cleaner //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test
#include "third_party/tensorflow/core/lib/io/path.h"
         ^
third_party/tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc:16:10: error: module //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/platform/env.h'
see http://go/cpp-features#layering_check; to fix run:
build_cleaner //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test
#include "third_party/tensorflow/core/platform/env.h"
```

**Comment by frreiss** [comment]
> Internal tests fails with:
> 
> ```
> third_party/tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc:15:10: error: module //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/lib/io/path.h'
> see http://go/cpp-features#layering_check; to fix run:
> build_cleaner //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test
> #include "third_party/tensorflow/core/lib/io/path.h"
>          ^
> third_party/tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc:16:10: error: module //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test does not depend on a module exporting 'third_party/tensorflow/core/platform/env.h'
> see http://go/cpp-features#layering_check; to fix run:
> build_cleaner //third_party/tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test
> #include "third_party/tensorflow/core/platform/env.h"
> ```

Commit 0f2c182 adds additional direct dependencies to the test target that should fix these errors.

**Comment by frreiss** [comment]
> @frreiss `DatasetParams` and `DatasetOpsTestBaseV2` are updated to make it easier to build the dataset pipeline in the test. [Here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc) is an example. Could you please update the tests for `LMDBDatasetOp`? Feel free to let me know if there is anything I can help.

Thanks @feihugis , I've updated the test case to use the latest version of the test harness in commit ab26059.

**Review by rachellim — DISMISSED** [review]

**Comment by rachellim** [comment]
This is failing internal tests (click on Details > {TARGET} > Target Log). Can you fix? Thanks.

**Comment by frreiss** [comment]
Updated this branch so the code works with the latest version of the test harness.

**Review by rachellim — APPROVED** [review]

**Review by rachellim — DISMISSED** [review]

**Comment by gbaned** [comment]
@frreiss Could you please check failed build errors? Thanks!

**Comment by frreiss** [comment]
One of the member functions of the test harness changed names. I've updated my code accordingly. Tests should pass now.

**Review by rachellim — DISMISSED** [review]

**Comment by gbaned** [comment]
@frreiss here is the internal error we are getting, can you please check this

> third_party/tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc:84
Expected equality of these values:
  ::tensorflow::Status::OK()
    Which is: OK
  (fs->CopyFile(src_loc, dest_loc))
    Which is: Not found: open failed for tensorflow/core/lib/lmdb/testdata/data.mdb: No such file or directory
Stack trace:
0x7f6212397803: tensorflow::data::experimental::(anonymous namespace)::MaybeCopyDataFile() @ ??:??
0x7f6212397ce1: tensorflow::data::experimental::(anonymous namespace)::LMDBDatasetOpTest_InvalidPathAtStart_Test::TestBody() @ ??:??
0x7f620f0befe9: testing::Test::Run() @ ??:??
0x7f620f0c016f: testing::TestInfo::Run() @ ??:??

cc @rachellim @jsimsa

**Comment by frreiss** [comment]
That is a strange failure. I'm not able to reproduce that behavior on my Mac and Linux test machines. On both machines, `bazel test //tensorflow/core/kernels/data/experimental:lmdb_dataset_op_test` succeeds in the running the test case.

@gbaned What command is your internal build using to invoke the test case?

**Comment by rachellim** [comment]
You should use `testing::TensorFlowSrcRoot()` instead of the `tensorflow` path directly, i.e.

```string src_loc = io::JoinPath(testing::TensorFlowSrcRoot(), kDataFileLoc, kDataFileName);```

**Review by rachellim — DISMISSED** [review]

File: tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    return LMDBDatasetOp::kDatasetType;
+  }
+
+ private:
+  // Names of binary database files to read, boxed up inside a Tensor of
+  // strings
+  Tensor filenames_;
+};
+
+class LMDBDatasetOpTest : public DatasetOpsTestBaseV2 {};
```
Comment (feihugis) [inline tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc]: @frreiss As `DatasetOpsTestBaseV2` has been merged into `DatasetOpsTestBase`, could you please change `DatasetOpsTestBaseV2` to `DatasetOpsTestBase`?

**Comment by gbaned** [comment]
@frreiss Could you please address feihugis's comments and keep us posted. Thanks!

File: tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc (hunk 2 above)
Comment (frreiss) [reply tensorflow/core/kernels/data/experimental/lmdb_dataset_op_test.cc]: Changed the base class to `DatasetOpsTestBase` in commit ca5e8a2.

**Review by feihugis — DISMISSED** [review]

**Review by rachellim — DISMISSED** [review]

**Comment by jsimsa** [comment]
The internal tests are indicating the following memory leak:

```
E0103 09:55:04.665212    7665 heap-profile-table.cc:445] RAW: Leak check _main_ detected leaks of 464 bytes in 2 objects
E0103 09:55:04.665305    7665 heap-profile-table.cc:465] RAW: The 2 largest leaks:
E0103 09:55:04.678486    7665 heap-profile-table.cc:486] RAW: Leak of 232 bytes in 1 objects allocated from:
	@ 0x7f5fd04e76d9 mdb_env_create
	@ 0x5639e5c1ef10 tensorflow::data::experimental::LMDBDatasetOp::Dataset::Iterator::SetupStreamsLocked()
	@ 0x5639e5c1e470 tensorflow::data::experimental::LMDBDatasetOp::Dataset::Iterator::GetNextInternal()
	@ 0x7f5fceb9d515 tensorflow::data::DatasetBaseIterator::GetNext()
	@ 0x7f5fd051053e tensorflow::data::experimental::(anonymous namespace)::LMDBDatasetOpTest_InvalidPathAtStart_Test::TestBody()
	@ 0x7f5fcca0c1ca testing::Test::Run()
	@ 0x7f5fcca0d330 testing::TestInfo::Run()
	@ 0x7f5fcca0dd27 testing::TestSuite::Run()
	@ 0x7f5fcca1d187 testing::internal::UnitTestImpl::RunAllTests()
	@ 0x7f5fcca1c795 testing::UnitTest::Run()
	@ 0x7f5fd04cdd2e main
	@ 0x7f5fc6d18bbd __libc_start_main
	@ 0x5639e5c1c769 ../sysdeps/x86_64/start.S:121 _start

E0103 09:55:04.678629    7665 heap-profile-table.cc:486] RAW: Leak of 232 bytes in 1 objects allocated from:
	@ 0x7f5fd04e76d9 mdb_env_create
	@ 0x5639e5c1ef10 tensorflow::data::experimental::LMDBDatasetOp::Dataset::Iterator::SetupStreamsLocked()
	@ 0x5639e5c1e470 tensorflow::data::experimental::LMDBDatasetOp::Dataset::Iterator::GetNextInternal()
	@ 0x7f5fceb9d515 tensorflow::data::DatasetBaseIterator::GetNext()
	@ 0x7f5fd051125b tensorflow::data::experimental::(anonymous namespace)::LMDBDatasetOpTest_InvalidPathInMiddle_Test::TestBody()
	@ 0x7f5fcca0c1ca testing::Test::Run()
	@ 0x7f5fcca0d330 testing::TestInfo::Run()
	@ 0x7f5fcca0dd27 testing::TestSuite::Run()
	@ 0x7f5fcca1d187 testing::internal::UnitTestImpl::RunAllTests()
	@ 0x7f5fcca1c795 testing::UnitTest::Run()
	@ 0x7f5fd04cdd2e main
	@ 0x7f5fc6d18bbd __libc_start_main
	@ 0x5639e5c1c769 ../sysdeps/x86_64/start.S:121 _start
```

**Comment by gbaned** [comment]
@frreiss Can you please check jsimsa's comments and keep us posted? Thanks!

**Comment by frreiss** [comment]
@jsimsa it looks like `LMDBDataset.Iterator` leaks connections to the LMDB database. I've added some fixes to this branch.

I've also made some changes to conform to the new way of creating string tensors.

**Review by jsimsa — APPROVED** [review]

**Review by gbaned — APPROVED** [review]

**Pull request merged**